plugins/code-climate/src/api/production-api.ts

```diff
@@ -0,0 +1,184 @@
+
+const basicIssuesQuery = `${encodeURI('page[size]')}=1&${encodeURI(
+  'filter[status][$in][]',
+)}=open&${encodeURI('filter[status][$in][]')}=confirmed&${encodeURI(
+  'page[size]',
+)}=1&${encodeURI('filter[categories][$in][]')}=`;
+
+const codeSmellsQuery = `${basicIssuesQuery}Complexity`;
+
+const duplicationQuery = `${basicIssuesQuery}Duplication`;
+
+const otherIssuesQuery = `${basicIssuesQuery}Bug Risk`;
+
+export class ProductionCodeClimateApi implements CodeClimateApi {
+  constructor(
+    private readonly discoveryApi: DiscoveryApi,
+    private readonly identityApi?: IdentityApi,
+  ) {}
+
+  async fetchAllData(
+    apiUrl: string,
+    options: any,
+    repoID: string,
+    snapshotID: string,
+    testReportID: string,
+  ): Promise<any> {
+    const [
+      maintainabilityResponse,
+      testCoverageResponse,
+      codeSmellsResponse,
+      duplicationResponse,
+      otherIssuesResponse,
+    ] = await Promise.all([
+      await fetch(`${apiUrl}/repos/${repoID}/snapshots/${snapshotID}`, options),
+      await fetch(
+        `${apiUrl}/repos/${repoID}/test_reports/${testReportID}`,
+        options,
+      ),
+      await fetch(
+        `${apiUrl}/repos/${repoID}/snapshots/${snapshotID}/issues?${codeSmellsQuery}`,
+        options,
+      ),
+      await fetch(
+        `${apiUrl}/repos/${repoID}/snapshots/${snapshotID}/issues?${duplicationQuery}`,
+        options,
+      ),
+      await fetch(
+        `${apiUrl}/repos/${repoID}/snapshots/${snapshotID}/issues?${otherIssuesQuery}`,
+        options,
+      ),
+    ]);
+
+    if (
+      (maintainabilityResponse.status >= 400 &&
+        maintainabilityResponse.status < 600) ||
+      (testCoverageResponse.status >= 400 &&
+        testCoverageResponse.status < 600) ||
+      (codeSmellsResponse.status >= 400 && codeSmellsResponse.status < 600) ||
+      (duplicationResponse.status >= 400 && duplicationResponse.status < 600) ||
+      (otherIssuesResponse.status >= 400 && otherIssuesResponse.status < 600)
+    ) {
+      throw new Error('Failed fetching Code Climate info');
+    }
+
+    const maintainabilityData = (
+      (await maintainabilityResponse.json()) as CodeClimateMaintainabilityData
+    ).data.attributes.ratings[0];
+    const testCoverageData = (
+      (await testCoverageResponse.json()) as CodeClimateTestCoverageData
+    ).data.attributes.rating;
+    const codeSmellsData = (
+      (await codeSmellsResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+    const duplicationData = (
+      (await duplicationResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+    const otherIssuesData = (
+      (await otherIssuesResponse.json()) as CodeClimateIssuesData
+    ).meta.total_count;
+
+    return [
+      maintainabilityData,
+      testCoverageData,
+      codeSmellsData,
+      duplicationData,
+      otherIssuesData,
+    ];
+  }
+
+  async fetchData(repoID: string): Promise<CodeClimateData | null> {
```

Comment:
Always avoid null at all cost, use undefined instead. But why does it support the empty repoID case at all? Could it not throw an error then instead, and force the caller to handle things instead?